File: clients/metorial-go/v1/endpoints/management/instance.go
package management

import (
	"github.com/metorial/metorial-go/v1/internal/endpoint"
	"github.com/metorial/metorial-go/v1/resources/instance"
)

// InstanceEndpoint provides access to an instance is an isolated environment within a Metorial project. Instances are created via the dashboard (since API keys are scoped to instances). Common setups include production, staging, and development instances.
type InstanceEndpoint struct {
	client *endpoint.Client
}

// NewInstanceEndpoint creates a new InstanceEndpoint.
func NewInstanceEndpoint(client *endpoint.Client) *InstanceEndpoint {
	return &InstanceEndpoint{client: client}
}

// Get retrieves metadata and configuration details for a specific instance.
func (e *InstanceEndpoint) Get(instanceId string) (*instance.InstanceGetOutput, error) {
	req := &endpoint.Request{
		Path: []string{"instances", instanceId, "instance"},
	}
	var result instance.InstanceGetOutput
	if err := e.client.Get(req, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: src/metorial/apis/core/src/controllers/instance/instance.ts
import { Controller } from '@metorial/rest';
import { checkAccess } from '../../middleware/checkAccess';
import { instanceGroup, instancePath } from '../../middleware/instanceGroup';
import { instancePresenter } from '@metorial/presenters';

export let instanceController = Controller.create(
  {
    name: 'Instance',
    description:
      'An instance is an isolated environment within a Metorial project. Instances are created via the dashboard (since API keys are scoped to instances). Common setups include production, staging, and development instances.',
    hideInDocs: true
  },
  {
    get: instanceGroup
      .get(instancePath('instance', 'instance.get'), {
        name: 'Get instance details',
        description: 'Retrieves metadata and configuration details for a specific instance.'
      })
      .use(checkAccess({ possibleScopes: ['organization.instance:read'] }))
      .output(instancePresenter)
      .do(async ctx => instancePresenter.present({ instance: ctx.instance }))
  }
);
